// @linaria/core/lib/processors/BaseProcessor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _getClassNameAndSlug = _interopRequireDefault(require("./utils/getClassNameAndSlug"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable class-methods-use-this */
class BaseProcessor {
  dependencies = [];
  interpolations = [];

  constructor(astService, params, tagExp, displayName, idx, options, context) {
    this.astService = astService;
    this.params = params;
    this.tagExp = tagExp;
    this.displayName = displayName;
    this.idx = idx;
    this.options = options;
    this.context = context;
    const {
      className,
      slug
    } = (0, _getClassNameAndSlug.default)(this.displayName, this.idx, this.options, this.context);
    this.className = className;
    this.slug = slug;
  }

  get lastInterpolation() {
    return this.interpolations[this.interpolations.length - 1];
  }
  /**
   * It is called for each resolved expression in a template literal.
   * @param node
   * @param source
   * @return chunk of CSS that should be added to extracted CSS
   */


}

exports.default = BaseProcessor;
